core/environment.ts

```diff
@@ -170,27 +191,42 @@ export class Environment {
       }
     }
 
-    const constructor = new Function(
-      "__env",
-      `return${sync ? "" : " async"} function __template (${dataVarname}) {
-        let __pos = 0;
-        try {
-          ${dataVarname} = Object.assign({}, __template.defaults, ${dataVarname});
-          const __exports = { content: "" };
-          ${code}
-          return __exports;
-        } catch (cause) {
-          throw __env.utils.templateError(__template, __pos, cause);
-        }
-      }
-      `,
-    );
-
-    const template: Template = constructor(this);
+    let template, context;
+    try {
+      const constructor = new Function(
+        "__env",
+        `return${sync ? "" : " async"} function __template(${dataVarname}) {
+          try {
+            ${dataVarname} = Object.assign({}, __template.defaults, ${dataVarname});
+            const __exports = { content: "" };
+            ${code}
+            return __exports;
+          } catch (cause) {
+            __env.utils.printRuntimeError(cause, __template.context);
+            throw cause;
+          }
+        }`,
+      );
+      template = constructor(this);
+      const body = constructor.toString();
+      context = { path, body, tokens, source };
+    } catch (syntaxError) {
+      // In synchronous contexts, we don't have "time" to figure out exactly
+      // what the issue is, so we just throw the raw error
+      if (sync) throw syntaxError;
+      if (!(syntaxError instanceof SyntaxError)) throw syntaxError;
+      context = { path, body: code, tokens, source };
+      const promise = printJSSyntaxError(syntaxError, context);
+      template = async function () {
+        await promise;
+        throw syntaxError;
+      };
+    }
     template.file = path;
     template.code = code;
     template.source = source;
     template.defaults = defaults || {};
+    template.context = context;
```

Comment:
What's the purpose of `template.context`? All this data (except tokens) is already stored in the template (in file, code and source properties).